src/main/xerus/monstercat/api/Playlist.kt

```diff
@@ -0,0 +1,87 @@
+package xerus.monstercat.api
+
+import javafx.collections.FXCollections
+import javafx.collections.ObservableList
+import mu.KotlinLogging
+import xerus.ktutil.javafx.properties.SimpleObservable
+import xerus.ktutil.javafx.properties.bindSoft
+import xerus.monstercat.api.response.Track
+import java.util.*
+import kotlin.collections.ArrayList
+
+object Playlist {
+	val logger = KotlinLogging.logger { }
+	
+	val tracks: ObservableList<Track> = FXCollections.observableArrayList()
+	val history = ArrayDeque<Track>()
+	val currentIndex = SimpleObservable<Int?>(null).apply {
+		bindSoft({
+			tracks.indexOf(Player.activeTrack.value).takeUnless { i -> i == -1 }
+		}, Player.activeTrack, tracks)
+	}
+	
+	var repeat = false
+	var shuffle = false
+	
+	operator fun get(index: Int): Track? = tracks.getOrNull(index)
+	
+	/** This property prevents songs that come from the [history] from being added to it again,
+	 * which would effectively create an infinite loop */
+	var lastPolled: Track? = null
+	
+	fun getPrev(): Track? = history.pollLast()?.also {
+		lastPolled = it
+		logger.debug("Polled $it from History")
+	} ?: currentIndex.value?.let { get(it - 1) }
+	
+	fun getNext() = when {
+		shuffle -> nextSongRandom()
+		repeat && (nextSong() == null) -> tracks[0]
+		else -> nextSong()
+	}
+	
+	fun addNext(track: Track) {
+		tracks.apply { remove(track); add(currentIndex.value?.let { it + 1 } ?: 0, track) }
+	}
+	
+	fun add(track: Track) {
+		tracks.apply { remove(track); add(track) }
+	}
+	
+	fun removeAt(index: Int?) {
+		if (index != null) tracks.removeAt(index)
+		else tracks.removeAt(tracks.size - 1)
+	}
+	
+	fun clear(){
+		history.clear()
+		tracks.clear()
+	}
+	
+	fun setTracks(playlist: Collection<Track>) {
+		history.clear()
+		tracks.setAll(playlist)
+	}
+	
+	fun nextSongRandom(): Track {
+		val index = (Math.random() * tracks.size).toInt()
+		return if(index == currentIndex.value && tracks.size > 1) nextSongRandom() else tracks[index]
```

Comment:
this if statement doesn't really make sense to me

Reply:
This checks if the randomly choosen song is the same as the last one, if so tries again. There is a failsaife is there is only one track, otherwise it would loop indefinitely.

Reply:
You should skip the random calculation directly if there is only one Track, and then calculate in a way that doesn't need recursion.

This would be my idea:
```kotlin
if(tracks.size == 1) 
	return tracks[0]
val advance = (Math.random() * tracks.size - 1).toInt()
return tracks[currentIndex.value.plus(advance).rem(tracks.size)]
```

Reply:
Found a quite simple way as well. b2ac140c6b05a632db5317ce9a9eb3b9f59477cb . We just pick a random number between 0 and size - 1 : if that number is equal or above the current index, we add 1. So we have ints that range from 0..lastIndex but skipping currentIndex !

Reply:
And there is a failsafe (takeUnless ?: 0) in case we have only one song.